mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSequenceGenerator.java

```diff
@@ -80,48 +109,96 @@ private void initIndex() {
      * @param inc The increment, Could be negative or positive. If 0, it is
      * assumed to be 1. Used only if the sequence does not exist prior to this
      * call
+     * @param poolSize If the sequence already has a pool associated
+     * with it, this is ignored, and the next id is used from the
+     * pool. Otherwise, a new pool with this size is initialized for
+     * the sequence
      *
      * If the sequence already exists, the <code>init</code> and
      * <code>inc</code> are ignored.
      *
      * @return The value of the sequence before the call
      */
-    public long getNextSequenceValue(String name, long init, long inc) {
+    public long getNextSequenceValue(String name, long init, long inc, long poolSize) {
         LOGGER.debug("getNextSequenceValue({})", name);
-        // Read the sequence document
-        BasicDBObject q = new BasicDBObject(NAME, name);
-        DBObject doc = coll.findOne(q,null,ReadPreference.primary());
-        if (doc == null) {
-            // Sequence document does not exist. Insert a new document using the init and inc
-            LOGGER.debug("inserting sequence record name={}, init={}, inc={}", name, init, inc);
-            if (inc == 0) {
-                inc = 1;
+        // First check if there is already a pool of ids available
+        String fullName=coll.getFullName()+"."+name;
+        rwl.readLock().lock();
+        SequenceInfo si=sequenceInfo.get(fullName);
+        rwl.readLock().unlock();
```

Comment:
Readers of the hashmap use readLock, writers use writeLock. So, many can read, only one can write. Readers will block when someone is writing, but readers won't block each other. Better than synchronized.